### DISCUSSION TITLE
use of SSD1306 display with an already initialized i2c (wire) object

### DISCUSSION THREAD

**Discussion by jmibk** [olikraus/u8g2#1951] [Q&A]
I try to connect a SSD1306 display to a relatively big project that uses already two I2C interfaces on an ESP32.
Now I like to add an SSD1306 display to I2C line 2, where also other devices are connected (with different i2c addresses).

Now I tried to find a consturctor that supports the second I2C line with custom pin mapping, but there isn't any.
Is it possible to pass an TwoWire object to the U8G2 library?

My I2C object looks like: TwoWire I2C_2 = TwoWire(1);

Now I tried to find a constructor like

U8G2_SSD1306_128X64_NONAME_F_HW_I2C u8g2(U8G2_R0, I2C_2); or something else.

Software I2C only for the display isn't an option, because speed and usage of additional pins on the ESP32 (I use already ethernet, CAN, SD card, two I2C lines, RS232 and some relays).

Is there a way to use an existing TwoWire object with this library?

**Comment by olikraus** [comment]
Well, it is not the first time I read that wish and I already tried to write a constructor, which accepts a two wire object, but I always come to the point that any implementation will consume a lot of effort and time on my end. At least in the near future I wont have time to implement this. Instead my suggestion is to just replace the object in the library itself. I mean, the library is installed in your local folder and you can just modify it according to your needs.

For the standard I2C constructor, you just need to replace the Wire object here:
https://github.com/olikraus/u8g2/blob/a6843a12e1a7388f32eac88b832c25310e99a508/cppsrc/U8x8lib.cpp#L1313-L1360

**Reply by Friskiefriskie** [reply]
I searched a long time to find this reply. With this fix, second HW I2C works like a champ :-)

**Reply by FriqueNFraque** [reply]
any chance you can post your modified code? I am fighting through this to use two 64x32s... I have literally been looking for this answer for years.

**Comment by 01GOD** [comment]
Can explain a little more about how to do that? 
Currently have 2 TOF sensors each on a TwoWire instance and those are THANKFULLY both functioning, although working on adding an SSD1306 OLED display. Tried simply adding another TwoWIre instance and putting the address of that in the OLED constructor, but that didn't seem to function. Seems as though it's maybe hijacking the I2C bus also. 
I should note that I did use the default 21 and 22 GPIO on the ESP32 DO_IT for that OLED display. Don't know if that matters. 
Thanks in advance for answering soon! Hoping to get this done to finish the display part of a project today or tomorrow.

**Comment by 01GOD** [comment]
For ESP32, is it basically recommended to edit the SSD1306 library to have specified ints for pin numbers in this line?: 
Wire.begin((int)u8x8->pins[U8X8_PIN_I2C_DATA] , u8x8->pins[U8X8_PIN_I2C_CLOCK]);

**Reply by olikraus** [reply]
The arguments for the  Wire.begin() are taken over from the u8g2.begin() constructor.

**Comment by guillaumechardin** [comment]
> Well, it is not the first time I read that wish and I already tried to write a constructor, which accepts a two wire object, but I always come to the point that any implementation will consume a lot of effort and time on my end. At least in the near future I wont have time to implement this. Instead my suggestion is to just replace the object in the library itself. I mean, the library is installed in your local folder and you can just modify it according to your needs.
> 
> For the standard I2C constructor, you just need to replace the Wire object here:
> 
> https://github.com/olikraus/u8g2/blob/a6843a12e1a7388f32eac88b832c25310e99a508/cppsrc/U8x8lib.cpp#L1313-L1360

Hello i'm facing the same issue than  @jmibk .
Can you clarify your answer :
> For the standard I2C constructor, you just need to replace the Wire object here:
You mean edit some reference of `Wire` by `I2C_2`  (that is a global var if I use the jmbibk example ?) :
```
#ifdef U8X8_HAVE_HW_I2C 
   switch(msg) 
   { 
     case U8X8_MSG_BYTE_SEND: 
       I2C_2.write((uint8_t *)arg_ptr, (int)arg_int); 
       break; 
     case U8X8_MSG_BYTE_INIT: 
       if ( u8x8->bus_clock == 0 ) 	/* issue 769 */ 
 	u8x8->bus_clock = u8x8->display_info->i2c_bus_clock_100kHz * 100000UL; 
 #if defined(ESP8266) || defined(ARDUINO_ARCH_ESP8266) || defined(ESP_PLATFORM) || defined(ARDUINO_ARCH_ESP32) 
       /* for ESP8266/ESP32, Wire.begin has two more arguments: clock and data */           
       if ( u8x8->pins[U8X8_PIN_I2C_CLOCK] != U8X8_PIN_NONE && u8x8->pins[U8X8_PIN_I2C_DATA] != U8X8_PIN_NONE ) 
       { 
 	// second argument for the wire lib is the clock pin. In u8g2, the first argument of the  clock pin in the clock/data pair 
 	//I2C_2.begin((int)u8x8->pins[U8X8_PIN_I2C_DATA] , u8x8->pins[U8X8_PIN_I2C_CLOCK]); 
 	I2C_2.begin();
       } 
       else 
       { 
 	I2C_2.begin(); 
       } 
 #else 
       I2C_2.begin(); 
 #endif 
```

If it is ok, does editing U8x8lib.cpp will correct u8g2 class usage since the file is U8x8lib. I'm not sure that u8g2 is an extend of u8x8

Hope i was clear.
Thanks
Guillaume
